src/lib/libast/misc/fallbacks.c

```diff
@@ -22,6 +25,65 @@ void do_not_use_this_fallback() {
     abort();
 }
 
+#if !_lib_eaccess
+#if _lib_euidaccess
+// System doesn't have eaccess() but does have the equivalent euidaccess() so use that.
+int eaccess(const char *pathname, int mode) {
+    return euidaccess(pathname, mode);
+}
+#elif _lib_faccessat
+// System doesn't have eaccess() but does have faccessat() so use that.
+int eaccess(const char *pathname, int mode) {
+    return faccessat(AT_FDCWD, pathname, mode, AT_EACCESS);
+}
+#else
+// The platform doesn't have eaccess(), euidaccess(), or faccessat() so we have to roll our own.
+// This may not be optimal in that it calls access() when it might be possible to return an answer
+// based on what we already know. Since this is a fallback function we hope not to use we're
+// striving for simplicity and clarity.
+int eaccess(const char *pathname, int mode) {
+    // In case the user passed bogus bits for the mode tell them we can't determine what they want.
+    if (mode & ~(X_OK | W_OK | R_OK)) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    int uid = getuid ();
+    int gid = getgid ();
+    int euid = geteuid ();
+    int egid = getegid ();
+
+    // If we aren't suid or sgid then we can just use access().
+    // Similarly if we're just checking if the file exists just use access().
+    if ((euid == uid && egid == gid) || mode == F_OK) return access(pathname, mode);
```

Comment:
if `mode == F_OK` then we don't need to call `access`, we have already checked file exists above and can return `0`.

Reply:
Sorry. I see we are checking it after this line, so this change is fine.